File: aoc24/src/day07.clj
(ns aoc24.src.day07
  (:require
   [aoc-tools :refer [read-input submit-answer]]
   [utils :refer [nmap]]
   [clojure.string :as str]))

(def input (read-input :test))
(def real-input (read-input))

(defn ||
  [x y]
  (read-string (str x y)))

(defn unconcat
  [result num]
  (let [result (str result)
        pattern (re-pattern (str num "$"))]
    (read-string (str/replace result pattern ""))))

(defn achievable?
  [[result & numbers]]
  (cond
    (not (int? result)) false
    (= (count numbers) 1)  (= (first numbers) result)
    :else (or
           (achievable? (conj (butlast numbers) (- result (last numbers))))
           (achievable? (conj (butlast numbers) (/ result (last numbers))))
           (and
            ;; false  ;; Uncomment for part 1
            (str/ends-with? (str result) (str (last numbers)))
            (achievable? (conj (butlast numbers) (unconcat result (last numbers))))))))

(->> real-input
     (map (partial re-seq #"\d+"))
     (nmap 2 read-string)
     (filter achievable?)
     (map first)
     (apply +)
     #_(submit-answer 2))

File: aoc24/python/day07.py
with open("aoc24/inputs/day07.txt") as f:
    lines = f.read().splitlines()


def parse(line):
    result, rest = line.split(": ")
    return int(result), [int(n) for n in rest.split(" ")]


def unconcat(result, num):
    return int(str(result)[: -len(str(num))])


def is_achievable(result, numbers, part):
    if isinstance(result, float) and not result.is_integer():
        return False

    result = int(result)

    if len(numbers) == 1:
        return result == numbers[0]

    return (
        is_achievable(result - numbers[-1], numbers[:-1], part)
        or is_achievable(result / numbers[-1], numbers[:-1], part)
        or (
            part == 2
            and (str(result).endswith(str(numbers[-1])))
            and is_achievable(unconcat(result, numbers[-1]), numbers[:-1], part)
        )
    )


part_1 = part_2 = 0

for line in lines:
    result, numbers = parse(line)
    if is_achievable(result, numbers, part=1):
        part_1 += result
        part_2 += result
    elif is_achievable(result, numbers, part=2):
        part_2 += result


print(f"Part 1: {part_1}")
print(f"Part 2: {part_2}")
